tests/phpunit/integration/Modules/Analytics_4Test.php

```diff
@@ -206,13 +237,848 @@ public function test_get_datapoints() {
+			array(
+				'propertyID' => $property_id,
+			)
+		);
+
+		$http_client = $this->create_fake_http_client( $property_id );
+		$this->analytics->get_client()->setHttpClient( $http_client );
+		$this->analytics->register();
+
+		$this->analytics->get_data(
+			'report',
+			array(
+				// Note, metrics is a required parameter.
+				'metrics'    => 'sessions',
+				'dimensions' => 'sessionDefaultChannelGrouping,pageTitle',
+			)
+		);
+
+		$request_params = $this->request_handler_calls[0]['params'];
+
+		$this->assertEquals(
+			array(
+				array(
+					'name' => 'sessionDefaultChannelGrouping',
+				),
+				array(
+					'name' => 'pageTitle',
+				),
+			),
+			$request_params['dimensions']
+		);
+	}
+
+	/**
+	 * @dataProvider data_access_token
+	 *
+	 * When an access token is provided, the user will be authenticated for the test.
+	 *
+	 * @param string $access_token Access token, or empty string if none.
+	 */
+	public function test_get_report__dimensions_as_single_object( $access_token ) {
+		$this->setup_user_authentication( $access_token );
+
+		$property_id = '123456789';
+
+		$this->analytics->get_settings()->merge(
+			array(
+				'propertyID' => $property_id,
+			)
+		);
+
+		$http_client = $this->create_fake_http_client( $property_id );
+		$this->analytics->get_client()->setHttpClient( $http_client );
+		$this->analytics->register();
+
+		$this->analytics->get_data(
+			'report',
+			array(
+				// Note, metrics is a required parameter.
+				'metrics'    => 'sessions',
+				'dimensions' => array(
+					'name' => 'pageTitle',
+				),
+			)
+		);
+
+		$request_params = $this->request_handler_calls[0]['params'];
+
+		$this->assertEquals(
+			array(
+				array(
+					'name' => 'pageTitle',
+				),
+			),
+			$request_params['dimensions']
+		);
+	}
+
+	/**
+	 * @dataProvider data_access_token
+	 *
+	 * When an access token is provided, the user will be authenticated for the test.
+	 *
+	 * @param string $access_token Access token, or empty string if none.
+	 */
+	public function test_report__no_metrics( $access_token ) {
+		$this->setup_user_authentication( $access_token );
+
+		$property_id = '123456789';
+
+		$this->analytics->get_settings()->merge(
+			array(
+				'propertyID' => $property_id,
+			)
+		);
+
+		$http_client = $this->create_fake_http_client( $property_id );
+		$this->analytics->get_client()->setHttpClient( $http_client );
+		$this->analytics->register();
+
+		$data = $this->analytics->get_data( 'report', array() );
+
+		$this->assertWPErrorWithMessage( 'Request parameter is empty: metrics.', $data );
+	}
+
+	public function test_report__metric_validation() {
+		$this->enable_feature( 'dashboardSharing' );
+
+		$this->context        = new Context( GOOGLESITEKIT_PLUGIN_MAIN_FILE );
+		$options              = new Options( $this->context );
+		$this->user           = $this->factory()->user->create_and_get( array( 'role' => 'administrator' ) );
+		$user_options         = new User_Options( $this->context, $this->user->ID );
```

Comment:
Hmm, I had recreated them all in response to your comment about enabling the feature at the start of the test. I think it's valid to keep these because the fact that these instantiations shouldn't be subject to feature flag changes implies knowledge of their implementation that we arguably should not assume from a testing perspective.

To be honest I think we're doing a lot of this sort of instantiation in test `set_up` methods and then enabling features in the tests elsewhere in the test codebase, so it would not be inconsistent to do so, but also think it's a reasonable call to leave this as-is for the reason outlined above. I've left them as they are.

Reply:
Sounds good to me 👍